3.0 Multiple Flash messages

HavokInspiration opened the pull request:
Add the ability to the FlashComponent and FlashHelper to set and render multiple flash messages.
I also implementend the ability to enable the stacking at a late state (see the tests).
The FlashComponent also got 2 new method : `delete()` and `clear()`. 

`delete()` allows the deletion of flash messages based on index if the stacking ability is enabled and the message if not
`clear()` allows the deleting of messages based on their types. About this method, I went with the current implementation. I'm not really sure if it's the best, performance wise. Another approach would be to keep a map of indexes to quickly get them and delete them. I'm open to suggestion on this.

The `set()` and `__call()` method now returns the index of the inserted message, or null if stacking is disabled.

To enable the stacking, I did not create a specific method, the call to `config()` being straightforward enough in my opinion. But I'm not against having a `enableStacking()` method.

It is also possible to have the stack of messages rendered in a specific element, that can be setup in the FlashHelper configuration.

More details on #4619

HavokInspiration committed:
The FlashComponent can stack registered message.

This gives the ability to users to display multiple flash messages
of different types in their applications. Various methods were added
to allow a greater flexibility in the messages management (delete messages,
clear all message from a given type, edit a message).
The FlashHelper was also modified to follow this changes. It can also rendered
stacked messages in a given element. The option is configurable through
the "stackElement" config option

dereuromark commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -91,12 +96,108 @@ public function set($message, array $options = [])
+        if ($this->config('stacking.enabled') === true) {
+            $messages = $this->_session->read($sessionKey);
+
+            if (is_null($messages)) {
```
Should be `=== null`

HavokInspiration replied:
Will do.

dereuromark commented on src/View/Helper/FlashHelper.php:
```diff
@@ -77,13 +89,54 @@ public function render($key = 'flash', array $options = [])
+    {
+        $out = '';
+        foreach ($messages as $message) {
+            $message = $options + $message;
```
Are you sure you want $options to have a higher merge priority?
Usually I would expect this to be `$message += $options;`. But maybe you want $options to overwrite any set key?

HavokInspiration replied:
Well actually, it was to follow the original options attribution in the ``render()`` method : https://github.com/cakephp/cakephp/blob/3.0/src/View/Helper/FlashHelper.php#L80

Since ``$options`` is the one passed to the render method, I figured it should be the same when the stack is involved, to have consistency in the rendering behavior.

dereuromark replied:
Sure. Maybe allowing to overwrite them at runtime makes sense after all.

markstory commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -91,12 +96,108 @@ public function set($message, array $options = [])
             'params' => $options['params']
         ]);
+
+        return isset($index) ? $index : null;
```
You could default index to null.

markstory commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -42,7 +42,11 @@ class FlashComponent extends Component
+        'params' => [],
+        'stacking' => [
+            'enabled' => false,
+            'limit' => 50
```
Why do you need a limit?

HavokInspiration replied:
I dis this to avoir session overflowing in case things go bas. But it can be removed

markstory commented on src/View/Helper/FlashHelper.php:
```diff
@@ -26,7 +26,16 @@ class FlashHelper extends Helper
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'stackElement' => false
```
How will the helper share config data with the component?

HavokInspiration replied:
I don't really understand the question. Why should they share data? This config option is a string that contain the path to an optionnal element in which the stack of messages can be rendered

markstory replied:
My concern was that the setting could be enabled in the component, but not in the helper by accident causing the helper output to break.

HavokInspiration replied:
Oh ok. Maybe the name of the config option is misleading ?
I also thought about being able to overwrite this with the ``$options`` array of the ``render()`` method, since every other parameters can be overwritten.

markstory commented:
I wonder if the config option for stacking is needed? Could set() just always append? We could also add methods for overwriting existing messages.

HavokInspiration commented:
I don't know if it is needed but I did this to avoid compatibility break. If we go with an array all the way through, app that access the messages directly in the session will have a change in the format of the data that is fetched. Maybe some app use the fact that only one message can be set and use this to overwrite the message in some cases. That would break without this.

But I can remove this. We could simply add a note in the migration guide about this behaviour.
Overwriting message could be done by accessing them with the index

markstory commented:
I think changing the internalish data structure for flash messages is a reasonable thing to do in a minor release.

HavokInspiration commented:
Ok, I'll work on this.

HavokInspiration commented:
I refactored this to have the FlashComponent always append.
I also added an ``edit()`` method.
I also added a new possible option for the ``set()`` method : "reset". When used in a ``set()`` call, it will reset the stack and add the new message in a newly indexed stack.
I also squashed my commits.

Should I open a new PR with the 3.1 branch as target (I already rebased my branch on it)

jadb commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -42,7 +43,9 @@ class FlashComponent extends Component
-        'params' => []
+        'params' => [],
+        'reset' => false,
+        'stackLimit' => 50
```
Should this be 'limit', like for the paginator options? I know they are unrelated limits but limits nonetheless, what does everything think?

HavokInspiration replied:
I'll change it. Consistency is a good thing.

ADmad commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -83,20 +88,160 @@ public function set($message, array $options = [])
+    public function edit($index, $message, $key = 'flash')
+    {
+        $sessionKey = 'Flash.' . $key . '.' . $index;
+        if ($this->_session->check($sessionKey) === false) {
```
`if (!$this->_session->check($sessionKey))`

jadb commented:
+1 for the great job on this one @HavokInspiration!

HavokInspiration commented:
Hm, after going through this, the ``$options`` argument of the ``FlashHelper::render()`` does not make much sense anymore, because all messages in the stack will get the new options if it is used (if new parameters are added or if the element is overwritten for instance).
I don't know if that's something we want...

HavokInspiration committed:
Comments

HavokInspiration committed:
Change the "stackLimit" config option to "limit"

HavokInspiration committed:
Change a === false test

HavokInspiration commented:
Hm, after going through this, the ``$options`` argument of the ``FlashHelper::render()`` does not make much sense anymore, because all messages in the stack will get the new options if it is used (if new parameters are added or if the ``element`` param is overwritten for instance).
I don't know if that's something we want...

What if we provided a callback ? That will be triggered on each message rendering ? Kind of like a ``beforeRender`` hook. This give the user a finer control over what could be overwritten or not.

Something like :

```php
 $this->Flash->render('flash', function ($message) {
        if ($message['params']['title'] === 'Some title') {
            // change something in the $message array
        }
        return $message;
 });
```
The example does not make much sense, but you get the idea...
Where ``$message`` is the Flash message array.

HavokInspiration committed:
Add a missing parameter in the docblock

markstory commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -83,20 +93,159 @@ public function set($message, array $options = [])
+            end($messages);
+            $index = key($messages);
+            reset($messages);
+            $index++;
```
`$index = count($messages)` ?

HavokInspiration replied:
I did this at first but that brought problems if we can delete specific messages. I we use count, this :

```php
$this->Flash->success('0');
$this->Flash->success('1');
$this->Flash->success('2');
$this->Flash->delete(1);
$this->Flash->success('3');
```

will output 0 and 3 because, after deletion, the count is 2 and we already have a message indexed at 2. Which is why I had to use this.
But since we are killing ``delete()`` and ``clean()``, ``count()`` can be back in the game.

markstory commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -83,20 +93,159 @@ public function set($message, array $options = [])
+     * @param string $key The flash key where the message is stored
+     * @return void
+     */
+    public function edit($index, $message, $key = 'flash')
```
Perhaps replace() would be a better name?

markstory commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -83,20 +93,159 @@ public function set($message, array $options = [])
+    {
+        $sessionKey = 'Flash.' . $key . '.' . $index;
+        if (!$this->_session->check($sessionKey)) {
+            return;
```
Is this an exceptional case? I also wonder if we shouldn't just set the new message at the given index.

markstory commented on src/View/Helper/FlashHelper.php:
```diff
@@ -77,9 +89,44 @@ public function render($key = 'flash', array $options = [])
+            $out .= $this->_render($message);
+        }
+
+        $stackElement = $this->config('stackElement');
```
Nice :+1:

markstory commented on tests/TestCase/Controller/Component/FlashComponentTest.php:
```diff
@@ -139,13 +269,211 @@ public function testSetWithComponentConfiguration()
+
+        $this->Flash->set('This is a test message');
+        $this->Flash->set('This is another test message');
+        $this->assertEquals(2, count($this->Session->read('Flash.flash')));
```
PHPUnit has `assertCount()`.

markstory commented on tests/TestCase/Controller/Component/FlashComponentTest.php:
```diff
@@ -139,13 +269,211 @@ public function testSetWithComponentConfiguration()
+    {
+        $this->assertNull($this->Session->read('Flash.flash'));
+        $this->Flash->set('This is a test message');
+        $this->Flash->clear('default');
```
Does '' also work here?

markstory commented on tests/TestCase/Controller/Component/FlashComponentTest.php:
```diff
@@ -190,4 +520,146 @@ public function testCall()
+        $result = $this->Session->read('Flash.flash');
+        $this->assertEquals($expected, $result);
+
+        $this->Flash->success('It worked too', ['reset' => true]);
```
The parameter could almost be named 'clear'

jadb replied:
:+1: for consistency sake (other places use the 'clear' method to reset static properties).

HavokInspiration replied:
I'll change it.

bcrowe commented:
A couple thoughts.

Why do we need public `edit()`, `delete()`, and `clear()` methods? Setting flash messages should be a very intentional thing. Going back on previous decisions and saying "nevermind" to a message (edit), destroying a key set (clear), or deleting a particular message (delete) -- sounds like potentially muddy controller logic to me.

I also don't see much value in setting a limit. Seems like too much of a safeguard. It's not like the `Session` class does any hand-holding in this regard anyways.

My original thoughts were to just keep pushing to the stack, render them, and let `FlashHelper`'s `render()` nuke it all. A simple and clean set and forget.

markstory commented:
@bcrowe I agree that having edit, delete and clear is somewhat redundant. We could skip out on all but clear/delete and see if there is a need for them later. It is always easier to add methods later than to try and remove them :smile:

ADmad commented:
I too agree with @bcrowe that the FlashComponent (and FlashHelper) should be clean and simple. Those who need more complex features should use their own component/helper.

dakota commented:
I'm going to throw my 2c in here as well. I too agree with @bcrowe. Clearing and deleting flash messages is already possible using the existing session handler (If a developer wishes to clear them). I'm also a bit hesitant on the limit for flash messages.

HavokInspiration commented:
Well, we seem to have a consensus on this.
I'll clean everything up tonight if I find the time (I'll backup my branch, just in case :D)

This only thing I saw a real value in was the ``clear()`` method (after an enhancement where ``$type`` can be an array) for cases like this :

```php
$this->Flash->success('Operation A ok');
$this->Flash->success('Operation B ok');
$this->Flash->info('Some note about operation B');

//something bad happend
$this->Flash->clear(['success', 'info']);
$this->Flash->error('Everything is broken');
```

That might be an edge case though, but this is something I'd see myself using.
However, that would only really work (and that comment also applies to the current state of the method) if the ``element`` param is not overridden since the method is currently using the last part of the ``element`` param to check the "type" of the message. This would work good for the above code sample but might be less easy with ``set()`` and some fancy ``element`` param configuration.
Unless we were to add a ``type`` key to the Flash messages array (where "default" would be the default.) 
Anyway, I stated above, that is something I can do by extending the Component and add the method in my apps.

I'll clean everything up and will keep you posted.

ADmad commented:
> Unless we were to add a type key to the Flash messages array (where "default" would be the default.) 

Yes, it would be better to add a `type` key.

HavokInspiration commented:
To sum up your feedback :

- Remove the limit
- Remove the ``edit()`` method
- Remove the ``clear()`` method
- Remove the ``delete()`` method
- Rename the ``reset`` param to ``clear`` (for the ``set()`` method)
- Make the ``set()`` and ``__call()`` methods not return an index since it's not needed anymore
- Update DocBlocks and Test Cases accordingly

HavokInspiration commented:
Closing in favor of #6161 to target the 3.1 branch
